### ISSUE TITLE
Watcher dies when BlockGetter restarts with a rich UTXO set

### ISSUE THREAD

**Issue by jbunce** [omisego/elixir-omg#672]
```
watcher-parity-0 watcher ** (stop) exited in: GenServer.call(OMG.DB.LevelDBServer, {:multi_update, [{:put, :last_convenience_deposit_processor_eth_height, 4321221}]}, 5000)
watcher-parity-0 watcher     ** (EXIT) time out
watcher-parity-0 watcher     (elixir) lib/gen_server.ex:989: GenServer.call/3
watcher-parity-0 watcher     (stdlib) timer.erl:166: :timer.tc/1
watcher-parity-0 watcher     (omg_db) lib/db.ex:33: OMG.DB.multi_update/2
watcher-parity-0 watcher     lib/ethereum_event_listener.ex:115: OMG.API.EthereumEventListener.sync_height/2
watcher-parity-0 watcher     lib/ethereum_event_listener.ex:90: OMG.API.EthereumEventListener.handle_info/2
watcher-parity-0 watcher     (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
watcher-parity-0 watcher     (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
watcher-parity-0 watcher     (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
watcher-parity-0 watcher Last message: :sync⋅
```

This eventually kills the Watcher:

```
watcher-parity-0 watcher 2019-05-08 04:22:45.300 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for challenges_responds_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.303 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for ife_exit_finalizer⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.306 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for convenience_deposit_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.308 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for piggyback_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.310 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for exit_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.312 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for competitor_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.314 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for exit_challenger⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.316 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for depositor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.318 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for exit_finalizer⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.320 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for in_flight_exit_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.322 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for piggyback_challenges_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.324 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for convenience_exit_processor⋅
watcher-parity-0 watcher 2019-05-08 04:22:45.331 [info] module=application_controller function=info_exited/3 ⋅Application omg_watcher exited: shutdown⋅
```

**Comment by pdobacz** [comment]
The root cause of these `OMG.DB.LevelDBServer` timeouts was lying elsewhere, but these exceptions that came after are a result of a cascade of events and inability for the Watcher's supervisor to recover as it should. Due partly to the size of the Ari UTXO set. This shouldn't be this vulnerable so it is worth looking into.

Synopsis seems to be:
  - (root cause) `Postgrex` timeouts while processing an insert causing `BlockGetter` to restart
  - `BlockGetter`'s restart causes `State` to restart too
  - unfortunately a restart of `State` causes lengthy blocking writes to `LevelDB`, but at the same time the Watcher is syncing so the `EthereumEventListeners` keep talking to it causing timeouts seen in the logs above :point_up:
  - the restarting spree that unfolds isn't tolerant enough to handle and Watcher stops :(

Taking the liberty to name the issue to this specific problem

**Comment by pnowosie** [comment]
Wasn’t it done @pthomalla ? Looks similar to #724

**Comment by achiurizo** [comment]
@pdobacz @pnowosie any updates on this?

**Comment by pdobacz** [comment]
I looked at #724 and #750 again. To answer the question above, those changes seem to only have addressed one case of the restart cycle that the Watcher go into, not the root cause.

I'm thinking about what could a sensible short-term solution be here. First I thought about the old idea to split into several buckets (e.g. LevelDB dbs and Rocks equivalent) for lighter stream reads (which are the problem here). But on 2nd thought that doesn't help since we have a single threaded LevelDBServer - when a big stream read (UTXOs for example) is going on, the DB is blocked and syncing will flop, which is what's happening.

I also vaguely recall that we counted on the -> RocksDB migration to help here - with concurrent reads and better stream reading support. Is this a viable trail to pick up, @InoMurko?

**Comment by pnowosie** [comment]
Here is a summary of the experiments @pdobacz and I did

First current status of RocksDB implementation was investigated. 
- already uses "Prefix iterators"
- loading ~3M utxo set tooks 11-14 seconds - which is too long as `event_listeners` tries to update it's processing height and timeouts

#1066 experimental implementation does not load utxo set at `State` init.
Basing on it we wanted to see how large utxo set impacts reading exits / in-flight exits which are also stored in RocksDB.

With a perftest following network was prepared on my workstation

Total blocks | Total txs | Total outputs | Unspent outputs 
-------------- | ---------- | ---------------- | --------------------
29| 1665000|        4660057 |  2995057 

Also Watcher's `BlockGetter` was modified to insert 1 fake `exit_info` and 1 fake `in_flight_exit_info` per block. After Watcher syncs another 21 fakes (for each exit) was inserted.

Experiment shows that quering DB for exits and IFEs is very fast (1-3 ms)

**Comment by pdobacz** [comment]
Thanks @pnowosie.

Also reposting some thoughts from Slack here:

RocksDB + lazy utxo loading + Task-based hotfix to metrics calculation + Fuse should IMHO ameliorate (my next fav English word, after 'immaterial') large UTXO set stuff

- RocksDB - done but not on Samrong
- lazy utxo loading - is a part of the hacky #1066 , is possible to be extracted out of there and cleaned
- Task-based hotfix - merged by Ino long time ago
- Fuse - todo - but was dubbed optional IIRC, for lavish UTXO sets (I can work on figuring out what "lavish" means now)

**Comment by pdobacz** [comment]
considering what we have here, I believe this was fixed in #1103. "Fuse" is irrelevant for restarts, so I'll edit it out from the above post. (It was only relevant to metric-calculation-related issues with large UTXO set, so I posted this here erroneously)
